### PR TITLE
Avoid commiting the selected item when a commit character was typed at the start of a non-empty filter span.

### LINKED ISSUE #5487: Strange completion behavior at beginning of identifier
With name ```original``` in scope, type ```or```.  Completion appears.  Move to beginning of identifier with arrow keys and type ```!``` (i.e. oops, I meant to negate the expression).

I'm not sure what the expected result is, but the actual result is ```original!or```.

### PR DESCRIPTION
**Pull request by diryboy**
Fixes #5487

### CODE REVIEW

**Comment by Pilchie** [comment]
Tag @dotnet/roslyn-ide @rchande

**Comment by Pilchie** [comment]
test vsi please
test eta please

**Comment by diryboy** [comment]
Ops, what was the failure?

**Comment by Pilchie** [comment]
@dotnet/roslyn-infrastructure @tannergooding - is this related to updating the Jenkins machines to Update 1?  Should we close and re-open?

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs [hunk 1]
```diff
@@ -357,12 +372,12 @@ private bool IsFilterCharacter(char ch)
...
         }
 
-        private string GetCurrentFilterText(Model model, CompletionItem selectedItem)
+        private string GetCurrentFilterText(Model model, CompletionItem selectedItem, bool stopBeforeCaret = true)
         {
             var textSnapshot = this.TextView.TextSnapshot;
             var viewSpan = model.GetSubjectBufferFilterSpanInViewBuffer(selectedItem.FilterSpan);
             var filterText = model.GetCurrentTextInSnapshot(
-                viewSpan, textSnapshot, GetCaretPointInViewBuffer());
+                viewSpan, textSnapshot, endPoint: stopBeforeCaret ? (int?)GetCaretPointInViewBuffer() : null);
```
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: I don't remember why we pass in the caret point at all--what tests break if you always pass `null`?

Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: Oh, right, that's because the `FilterSpan` is usually just the edit span the item would use if committed, which of course has little bearing on the span to check for filtering.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs [hunk 2]
```diff
@@ -327,12 +327,27 @@ private bool IsCommitCharacter(char ch)
...
 
             var filterText = GetCurrentFilterText(model, selectedItem);
+            var isCommitCharacter = GetCompletionRules().IsCommitCharacter(selectedItem, typedChar, filterText);
+
+            // When a character, even appears to be a commit character, was typed at the begining of the filter span when it was empty,
+            // it's most likely that the user just realized he/she forgot to do something, pressed Ctrl+Left and typed it.
+            // i.e. blah| -> |blah -> !|blah
+            // but not: A a = new | -> A a = new (|
+            // Hence we should not commit the selected item.
+            if (isCommitCharacter && filterText.Length == 1)
```
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: Why is the length of the FilterText 1? Is it the commit character?

Comment (diryboy) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: Yes

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs (hunk 1 above)
Comment (diryboy) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: I do want to know. In VS2013, if you move the caret to the left changing something then press commit character, you get the identifier replaced, it's useful to correct something:
`Mon|` -> `Mo|n` -> `Mi|n` -> `Mi(|n` -> `Min(|`
In VS2015 you get `Min(|n` in the last step because of this caret check. It's in the code so I thought this change was made intentionally? Personally prefer VS2013 behavior.

Comment (diryboy) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: OK, if there's no spelling mistake and you want to wrap the existing method with some other `F(` before `Min(`, the VS2013 behavior will make you disappointed.. While in VS2015, correct spelling can use `Enter`, wrap thing before can use commit character.

Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: I agree, I think it would be better to make that change than to make the commit character inoperable at the start of the word (since most of the bug filers were trying to do that).

There are a couple of approaches I could imagine, let me know if you're interested in tackling it.

Comment (diryboy) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: I haven't verify what was the behavior in VS2013 when trying to wrap something before existing method call, it may not make one disappointed.

Comment (diryboy) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: VS2013 do make the commit character inoperable at the start of the word.

Comment (diryboy) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: After playing with both VS2015 and VS2013, I find that:

- When Ctrl+J was fired at the beginning of word, they both have empty filter span, so wrapping functions works well in both versions of VS.
- When Ctrl+J was fired in the middle of word, they both have the word in filter span, VS2013 replace the word when commit character typed, if the caret is NOT at the beginning; while VS2015 always replace the span between the start of the word and caret with the selected item when you use commit character, BUT always replace the whole word when you hit Enter.

I think the VS2015 behavior is arguably better. Consider when you first type something at the beginning of a word but no completion list shows, e.g. you have the following methods:
```C#
int Foo()
{
    return |Foo();
}

int BarVeryLong(int a) { return 0; }
```
Then you type `ba`, no completion list here, instead VS guesses you want to generate a new function later. So you have to press Ctrl+J, but now, the filter span will be the whole `baFoo`. When you find the `BarVeryLong` in the list and type `(`, VS2015 give you `BarVeryLong(Foo();`, VS2013 give you `BarVeryLong(();`. For this case, VS2015 wins!

**Comment by Pilchie** [comment]
I was just looking at a VS Feedback report about a similar issue.  The user said:

> This involves the intellisense when the cursor is in the middle of a word. In previous versions of Visual Studio, VS would ignore all the characters to the right of the cursor when suggesting words. VS 2015 does not igore characters to the right of the cursor, so it does not provide a list of suggestions if you have any garbage characters.
> 
>  For example, if the cursor is at the ^ position 
> 
>  string.Form^Xxxxxx
> 
>  and you hit ctrl+space to bring up the list of choice:
> 
>  * VS 2015 will sometimes not bring up any suggestions, or if does bring up the list it does not highlight "Format"
> 
>  * Previous versions of VS will bring up the list of methods with "Format" hightlighted.
>

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs (hunk 2 above)
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: Instead of doing this, can you get the current filter span and just check to see if the caret is at the start position with a "commit character" as the typed char?

**Comment by CyrusNajmabadi** [comment]
@Pilchie  I was just about hte file that bug.  It was driving me nuts and i only started looking last night to see why the IDE felt 'off'.  This is definitely annoying as i often go into the middle of a word and try to 'complete' again using the prefix to filter down the list accordingly.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs (hunk 2 above)
Comment (diryboy) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: Let me try again.

**Comment by diryboy** [comment]
Interesting, @Pilchie and @CyrusNajmabadi, looks like the other bug you described can be fixed by inserting `caretPosition` as the last parameter into [this line](https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs#L119). Do you want to keep it separated or included in this PR?

**Comment by rchande** [comment]
@diryboy Let's do that in a separate PR. When I was testing this yesterday, I noticed that in VS 2013
```C#
ArgumentNullException|ArgumentNullException
```
Ctrl-J at the | does a soft selection[1] even though the substring through the | is an exact match--I think this is to make typing a new method name and wrapping easier. We'll probably want Roslyn to have that behavior too.

[1] Soft selection: A unfilled box is drawn around the selected item.

**Comment by diryboy** [comment]
@rchande If I remember correctly, this hard/soft selection happened randomly and I couldn't figure out the reasons behind.

So let's keep these two issues separated.

Can I has review on my new commit? Also eta vsi runs.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs [hunk 3]
```diff
@@ -357,12 +373,12 @@ private bool IsFilterCharacter(char ch)
             return GetCompletionRules().IsFilterCharacter(selectedItem, ch, filterText);
         }
 
-        private string GetCurrentFilterText(Model model, CompletionItem selectedItem)
+        private string GetCurrentFilterText(Model model, CompletionItem selectedItem, bool stopBeforeCaret = true)
```
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: Is the `stopBeforeCaret` thing used any more? Doesn't look like it.

Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: `GetCurrentFilterText` does almost exactly the same thing as your code above that tracks the edit span forward in the view--consider turning this into `GetCUrrentFilterTextAndViewSpan` and returning the view span as an out param?

Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: You'll also want the overload of `GetCurrentCurrentTextInSnapshot` that takes a TextSpan return the ViewSpan as well, if you do that.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs [hunk 4]
```diff
@@ -327,12 +327,28 @@ private bool IsCommitCharacter(char ch)
...
-            var filterText = GetCurrentFilterText(model, selectedItem);
+            var snapshot = this.TextView.TextSnapshot;
+            var editSpan = model.GetCurrentSpanInSnapshot(model.GetSubjectBufferFilterSpanInViewBuffer(selectedItem.FilterSpan), snapshot);
+            var spanStart = (int)editSpan.Start;
+            var caretPosition = (int)GetCaretPointInViewBuffer();
+
+            var filterText = snapshot.GetText(Span.FromBounds(spanStart, caretPosition));
+            var isCommitCharacter = GetCompletionRules().IsCommitCharacter(selectedItem, typedChar, filterText);
+
+            // When a character, even appears to be a commit character, was typed at the begining of the filter span when it was not empty,
```
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: "When a commit character"

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs (hunk 3 above)
Comment (diryboy) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: Good, I like it

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs [hunk 5]
```diff
@@ -327,12 +327,28 @@ private bool IsCommitCharacter(char ch)
...
+            // it's most likely that the user just realized he/she forgot to do something, pressed Ctrl+Left and typed it.
+            // i.e. blah| -> |blah -> !|blah
+            // We should not commit the selected item in the above case.
+            // But be careful we still want completion in the following case: A a = new | -> A a = new (|
+            if (isCommitCharacter && caretPosition == spanStart + 1 && editSpan.Length > 1)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: I'd like @jasonmalinowski to weigh in on this.  Normally we map things to the subject buffer and make our determinations there.  But in this code we map things up to the view buffer.  This seems very different from our normal way of handling positions, views, buffers, and positions.   I'd prefer we stick with how we generally do things unless there is a strong reason to use the view's buffer.  I'm curious what jason thinks.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs [hunk 6]
```diff
@@ -313,7 +313,7 @@ private bool IsTextualTriggerCharacter(ICompletionService completionService, cha
             return completionService.IsTriggerCharacter(previousPosition.Snapshot.AsText(), previousPosition, GetCompletionProviders(), options);
         }
 
-        private bool IsCommitCharacter(char ch)
+        private bool IsCommitCharacter(char typedChar)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: Would a better approach to be to not even call this in the below case?  Here's what i'm thinking:

if the caret is at hte start of the filter span, should we even have anything selected strongly in the completion list?  And if we don't have something strongly selected, we shouldn't commit.

So the preferred behavior falls out not by saying "no this is not a commit character", but rather "we couldn't even commit here if we wanted to because there's no item selected.

**Comment by CyrusNajmabadi** [comment]
just to make sure, this doesn't break the following does it:

```
x<dot><dot>
```

If we've preselected something after ```x.```, then the second dot should commit that item.  I'd only want this new behavior if we're at the start of some non-empty span.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs (hunk 6 above)
Comment (diryboy) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: Changing to soft selection sounds reasonable. User in that have a visual to predict what will the IDE do next.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs (hunk 5 above)
Comment (Pilchie) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: IIRC @kayleh changed this, because in some asp.net cases the subject buffer may be disconnected and then re-connected to the view buffer in the course of a single session. :frowning:

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs (hunk 6 above)
Comment (diryboy) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: Currently Controller_CaretPositionChanged.cs does not wait for model, do you mind me placing
```
// TODO(cyrusn): Find a way to allow the user to cancel out of this.
```
as you did in Controller_TypedChar.cs?

**Comment by diryboy** [comment]
Switched to soft selection as @CyrusNajmabadi suggested.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs (hunk 5 above)
Comment (kayle) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: I don't remember all the details, but there were some edge cases in Razor:

Hello @title. @name.

The periods are both commit characters, but are not part of the C# subject buffer. I think typing another character after the period puts it back in the C# buffer (e.g. @title.ToString()).

**Comment by diryboy** [comment]
ping?

**Comment by rchande** [comment]
Thanks, @diryboy, this LGTM.

:+1:

**Comment by rchande** [comment]
Anyone else want to review @dotnet/roslyn-ide

**Comment by davkean** [comment]
@dotnet/roslyn-ide  Please take another look

**Comment by davkean** [comment]
@dotnet-bot test eta please

**Comment by davkean** [comment]
@dotnet-bot test vsi please

**Comment by diryboy** [comment]
What was the failure?

**Comment by diryboy** [comment]
@davkean Could you please rerun the failed runs in case they're caused by flaky issue? I could not see the details myself.

**Comment by diryboy** [comment]
Help, please. I cannot see the result of failed vsi runs.

**Comment by davkean** [comment]
I'll rerun because the screenshots and logs have been since been deleted ([this will fix this](https://github.com/dotnet/roslyn/issues/8553)), but P2/P3 failed with lots REPL failures like:
```
 Scenario: 4) Verify 'Execute Input'
    Executing action [3/1/2016 3:23:16 AM] SubmitReplText
    Executing action [3/1/2016 3:23:17 AM] VerifyReplOutputContains
 Scenario: 5) Verify 'Force New Line and Indent'
    Executing action [3/1/2016 3:23:17 AM] SetReplText
--------------------------------
System.Exception: Command Edit.SelectionCancel was not available after retrying for a timeout of 60000ms
   at Tao.VS.VisualStudioAction.ExecuteCommand(String commandName, String argument) in d:\j\workspace\dotnet_roslyn---b84db72c\Closed\Hosting\Test\Actions\VisualStudio\VisualStudioAction.cs:line 99
   at Roslyn.VSHosting.Test.SetReplTextAction.Execute(TaoEngine engine) in d:\j\workspace\dotnet_roslyn---b84db72c\Closed\Hosting\RoslynTaoActions\Actions\VisualStudio\SetReplTextAction.cs:line 50
   at Tao.Engine.DebuggableActionExecutor.Execute() in f:\dd\src\Omni\Tao\Tao.Engine\ActionExecutor.cs:line 139
   at Tao.Engine.DefaultActionExecutor.Execute() in f:\dd\src\Omni\Tao\Tao.Engine\ActionExecutor.cs:line 36
```

@dotnet-bot test p2 please
@dotnet-bot test p3 please

**Comment by davkean** [comment]
@dotnet-bot retest prtest/win/vsi/p2 please
@dotnet-bot retest prtest/win/vsi/p3 please

**Comment by diryboy** [comment]
Thanks @davkean, this time they don't fail. I don't think `SetReplText` was affected by this pr.

**Comment by davkean** [comment]
Can we have one more @dotnet/roslyn-ide sign off on this? @DustinCampbell @jasonmalinowski @Pilchie @CyrusNajmabadi in particular?

**Comment by diryboy** [comment]
Ping? Would like to @CyrusNajmabadi because of the idea, and the similar TODOs found in other files.

**Comment by davkean** [comment]
tag @dotnet/roslyn-ide

**Comment by Pilchie** [comment]
Sorry for the way too long delay.  :+1:

**Comment by diryboy** [comment]
Thanks, so can this get in?

**Comment by davkean** [comment]
@diryboy Yep, sorry about that. Thanks again Paul for the contribution!
